### PR TITLE
feat: watch search enhancement

### PR DESCRIPTION
**Pull request by edmonday**
# Description

Improve watch search enhancement by leveraging Algolia Search.

[- Link to Basecamp Todo](https://3.basecamp.com/3105655/buckets/36562320/todos/7195706912)

# How should this PR be QA Tested?

Please describe the QA tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

# Walkthrough

copilot:walkthrough

### CODE REVIEW

**Review by GeronimoJohn — APPROVED** [review]

**Review by GeronimoJohn — APPROVED** [review]

File: apps/watch/src/components/VideosPage/VideosPage.spec.tsx [hunk 1]
```diff
@@ -231,88 +163,68 @@ describe('VideosPage', () => {
...
       )
-      expect(push).toHaveBeenCalledWith('/videos?title=JESUS', undefined, {
-        shallow: true
-      })
+      expect(getByRole('heading', { name: 'The Savior' })).toBeInTheDocument()
+      expect(getByRole('button', { name: 'No More Videos' })).toBeDisabled()
     })
   })
+
+  // TODO: add test for load more button
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/VideosPage.spec.tsx:229]: Add this test

File: apps/watch/src/components/VideosPage/utils/getQueryParameters/getQueryParameters.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  // we intentionally use window.location.search to prevent multiple renders
+  // which occurs when using const { query } = useRouter()
+  const searchString =
+    typeof window !== 'undefined' ? window.location.search : undefined
+
+  function getFilter(): VideoPageFilter {
+    const query = new URLSearchParams(searchString?.split('?')[1])
+
+    const getQueryParamArray = (param: string | null): string[] | undefined =>
+      param != null ? [param] : undefined
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/utils/getQueryParameters/getQueryParameters.ts:17]: convert arrow function to normal function

File: apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+export function checkFilterApplied(filter: VideoPageFilter): boolean {
+  const filterToCheck = [
+    filter.title,
+    filter.availableVariantLanguageIds,
+    filter.subtitleLanguageIds
+  ]
+
+  return filterToCheck.some(
+    (value) => value !== undefined && value != null && value !== ''
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts:11]:
```suggestion
    (value) => value != null && value !== ''
```

File: apps/watch/src/components/VideosPage/VideosPage.tsx [hunk 4]
```diff
@@ -44,108 +31,58 @@ export const GET_LANGUAGES = gql`
...
-  const [previousCount, setPreviousCount] = useState(0)
 
-  // we intentionally use window.location.search to prevent multiple renders
-  // which occurs when using const { query } = useRouter()
-  const query = new URLSearchParams(
-    typeof window !== 'undefined'
-      ? window.location.search.split('?')[1]
-      : undefined
-  )
+  const localVideos = videos.filter((video) => video !== null)
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/VideosPage.tsx:41]:
```suggestion
  const localVideos = videos.filter((video) => video != null)
```

File: apps/watch/src/components/VideosPage/utils/useVideoSearch/useVideoSearch.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        const newHits = isEqual(previousFilter, filter)
+          ? [...hits, ...resultHits]
+          : resultHits
+
+        setHits(newHits)
+        setCurrentPage(pageNumber)
+        setTotalPages(totalPages)
+      } catch (error) {
+        console.error('Error occurred while searching:', error)
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/utils/useVideoSearch/useVideoSearch.ts:97]: We don't want to display an error message to user as well?

File: apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts (hunk 3 above)
Comment (edmonday) [reply apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts:11]: It's actually never null and instead undefined. Will update this

File: apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { VideoPageFilter } from '../getQueryParameters'
+
+export function checkFilterApplied(filter: VideoPageFilter): boolean {
+  const filterToCheck = [
+    filter.title,
+    filter.availableVariantLanguageIds,
+    filter.subtitleLanguageIds
+  ]
+
+  return filterToCheck.some((value) => value !== undefined && value !== '')
```
Comment (Kneesal) [inline apps/watch/src/components/VideosPage/utils/checkFilterApplied/checkFilterApplied.ts:10]:
```suggestion
  return filterToCheck.some((value) => value != null && value !== '')
```

**Review by Kneesal — APPROVED** [review]

**Pull request merged**